state/volume.go

```diff
@@ -266,7 +275,19 @@ func (st *State) addVolumeOp(params VolumeParams) (txn.Op, names.VolumeTag, erro
 	if err := st.validateVolumeParams(params); err != nil {
 		return txn.Op{}, names.VolumeTag{}, errors.Annotate(err, "validating volume params")
 	}
-	name, err := newVolumeName(st)
+	_, provider, err := poolStorageProvider(st, params.Pool)
```

Comment:
This validation code should be in validateVolumeParams I think. Modify the method signature to take in a machineId. Also, if scope is Environ and machineId != "", the current code sets machineId to "" but should we error instead?